source/grid/grid_out.cc

```diff
@@ -3219,28 +3256,39 @@ namespace internal
                   << '\n';
             }
           else
-            // cell is at boundary and we
-            // are to treat curved
-            // boundaries. so loop over
-            // all faces and draw them as
-            // small pieces of lines
+            // cell is at boundary and we are to treat curved boundaries. so
+            // loop over all faces and draw them as small pieces of lines
             {
               for (unsigned int face_no=0;
                    face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
                 {
+                  point_cache.clear();
                   const typename dealii::Triangulation<dim,spacedim>::face_iterator
                   face = cell->face(face_no);
-                  if (face->at_boundary() || gnuplot_flags.curved_inner_cells)
+                  if (dim != spacedim || face->at_boundary() || gnuplot_flags.curved_inner_cells)
                     {
-                      // compute offset
-                      // of quadrature
-                      // points within
-                      // set of projected
-                      // points
+                      // compute offset of quadrature points within set of
+                      // projected points
                       const unsigned int offset=face_no*n_points;
-                      for (unsigned int i=0; i<n_points; ++i)
-                        out << (mapping->transform_unit_to_real_cell
-                                (cell, q_projector->point(offset+i)))
+                      point_cache.push_back(face->vertex(0));
+                      for (unsigned int i=1; i<n_points - 1; ++i)
+                        point_cache.push_back(mapping->transform_unit_to_real_cell
+                                              (cell, q_projector->point(offset+i)));
+                      point_cache.push_back(face->vertex(1));
+                      while (point_cache.size() > 2)
+                        {
+                          Tensor<1, spacedim> first_difference = point_cache[1] - point_cache[0];
+                          first_difference /= first_difference.norm();
+                          Tensor<1, spacedim> second_difference = point_cache[2] - point_cache[1];
+                          second_difference /= second_difference.norm();
+                          // If the three points are colinear then remove the middle one.
+                          if ((first_difference - second_difference).norm() < 1e-10)
```

Comment:
You'll want this to be a relative tolerance, not an absolute one.

Reply:
I think this is okay: the differences are normalized to length 1 above.